tensorflow_addons/optimizers/novograd.py

```diff
@@ -0,0 +1,165 @@
+        for var in var_list:
+            self.add_slot(var=var, slot_name='m', initializer='zeros')
+        for var in var_list:
+            self.add_slot(
+                var=var,
+                slot_name='v',
+                initializer=tf.zeros(shape=[], dtype=var.dtype))
+
+    def _prepare_local(self, var_device, var_dtype, apply_state):
+        super(Novograd, self)._prepare_local(var_device, var_dtype,
+                                             apply_state)
+        local_step = tf.cast(self.iterations + 1, var_dtype)
+        beta_1_t = tf.identity(self._get_hyper('beta_1', var_dtype))
+        beta_2_t = tf.identity(self._get_hyper('beta_2', var_dtype))
+        beta_1_power = tf.pow(beta_1_t, local_step)
+        beta_2_power = tf.pow(beta_2_t, local_step)
+        lr = (apply_state[(var_device, var_dtype)]['lr_t'] *
+              (tf.sqrt(1 - beta_2_power) / (1 - beta_1_power)))
+        apply_state[(var_device, var_dtype)].update(
+            dict(
+                lr=lr,
+                epsilon=tf.convert_to_tensor(self.epsilon, var_dtype),
+                beta_1_t=beta_1_t,
+                beta_2_t=beta_2_t,
+                one_minus_beta_2_t=1 - beta_2_t,
+                one_minus_beta_1_t=1 - beta_1_t,
+            ))
+
+    def set_weights(self, weights):
+        params = self.weights
+        # If the weights are generated by Keras V1 optimizer, it includes vhats
+        # even without amsgrad, i.e, V1 optimizer has 3x + 1 variables, while V2
+        # optimizer has 2x + 1 variables. Filter vhats out for compatibility.
+        num_vars = int((len(params) - 1) / 2)
+        if len(weights) == 3 * num_vars + 1:
+            weights = weights[:len(params)]
+        super(Novograd, self).set_weights(weights)
+
+    def _resource_apply_dense(self, grad, var, apply_state=None):
+        var_device, var_dtype = var.device, var.dtype.base_dtype
+        coefficients = ((apply_state or {}).get((var_device, var_dtype))
+                        or self._fallback_apply_state(var_device, var_dtype))
+        weight_decay = self._get_hyper('weight_decay')
+        grad_averaging = self._get_hyper('grad_averaging')
+
+        v = self.get_slot(var, 'v')
+        g_2 = tf.reduce_sum(tf.square(tf.cast(grad, tf.float32)))
+        v_t = tf.cond(tf.equal(self.iterations, 0),
+                      lambda: g_2,
+                      lambda: v * coefficients['beta_2_t'] + g_2 * coefficients['one_minus_beta_2_t'])
+        v_t = v.assign(v_t, use_locking=self._use_locking)
+
+        grad = grad / (tf.sqrt(v_t) + self.epsilon)
+        grad = tf.cond(grad_averaging,
+                       lambda: grad * coefficients['one_minus_beta_1_t'],
+                       lambda: grad)
+        grad = tf.cond(tf.greater(weight_decay, 0),
+                       lambda: grad + weight_decay * var,
+                       lambda: grad)
+        m = self.get_slot(var, 'm')
+        return training_ops.resource_apply_momentum(
```

Comment:
I would have liked to use Adam here but from what I understand of Novograd it doesn't always use grad_averaging which makes it inconvenient to use Adam.